modules/tests/src/main/scala/Person.scala

```diff
@@ -17,43 +19,56 @@ object Person {
   val personToTupleIso = Iso[Person, (Seq[Char], Option[Role])](f)(g)
 }
 
-trait TestModule extends SchemaModule[JsonSchema.type] {
+trait TestModule extends SchemaModule[JsonSchema.type] with Versioning[JsonSchema.type] {
   val R = JsonSchema
 
   type PersonTuple = (Seq[Char], Option[Role])
 
-  val user = record(
-    "active" -*>: prim(JsonSchema.JsonBool) :*: "boss" -*>: self(person),
-    Iso[(Boolean, Person), User]((User.apply _).tupled)(u => (u.active, u.boss))
-  )
-
-  val admin = record(
-    "rights" -*>: seq(prim(JsonSchema.JsonString)),
-    Iso[List[String], Admin](Admin.apply)(_.rights)
-  )
-
-  val role = union(
-    "user" -+>: user :+:
-      "admin" -+>: admin,
-    Iso[User \/ Admin, Role] {
-      case -\/(u) => u
-      case \/-(a) => a
-    } {
-      case u @ User(_, _) => -\/(u)
-      case a @ Admin(_)   => \/-(a)
-    }
-  )
-
-  def person: Schema[Person] = record(
-    "name" -*>: prim(JsonSchema.JsonString) :*:
-      "role" -*>: optional(
-      role
-    ),
-    Iso[(String, Option[Role]), Person]((Person.apply _).tupled)(p => (p.name, p.role))
-  )
-
-  val personTupleSchema = iso[Person, PersonTuple](
-    person,
-    Person.personToTupleIso
-  )
+  val current = Current
+    .schema(
+      record(
+        "active".narrow -*>: prim(JsonSchema.JsonBool),
+        Iso[Boolean, User](User.apply)(u => u.active)
+      )
+    )
+    .schema(
+      record(
+        "rights".narrow -*>: seq(prim(JsonSchema.JsonString)),
+        Iso[List[String], Admin](Admin.apply)(_.rights)
+      )
+    )
+    .schema(
+      (u: Schema[User], a: Schema[Admin]) =>
+        union(
+          "user".narrow -+>: u :+:
+            "admin".narrow -+>: a,
+          Iso[User \/ Admin, Role] {
+            case -\/(u) => u
+            case \/-(a) => a
+          } {
+            case u @ User(_)  => -\/(u)
+            case a @ Admin(_) => \/-(a)
+          }
+        )
+    )
+    .schema(
+      (r: Schema[Role]) =>
+        record(
+          "name".narrow -*>: prim(JsonSchema.JsonString) :*:
```

Comment:
The `.narrow` is a little noticeable. No way to eliminate that?

Overall, nice clean syntax! ❤️

Reply:
It is indeed noticeable, but it gets worse...

The core doesn't know about singleton types (it doesn't even know about shapeless!). It only mandates that you use a type that is at least as specific as `R.ProductTermId` to label your fields. I the user forgets the `.narrow` part (or any other way to narrow the label's type), the above code will still compile, problems will only arise when they attempt a migration on that schema.

I haven't tried to lift that concern up to the core yet, but I see only two paths toward that:
* make the core depend on shapeless
* upgrade the core to Scala 2.13 (which has singleton types, AFAIK)

Both come with their drawbacks: we want zero dependency in core, but do we want to restrict ourselves to Scala 2.13?